File: ProED/JavaScript/proed.js
const ffi = require('ffi-napi');
const ref = require('ref-napi');

// --- DATENTYPEN ---
const uint8 = ref.types.uint8;
const uint32 = ref.types.uint32;
const size_t = ref.types.size_t;
const pointer = ref.refType(ref.types.void);

// --- KONSTANTEN ---
const HEADER_SIZE = 48; // 16 Byte IV + 32 Byte HMAC
const KEY_SIZE = 32;    // 256-Bit Key
const CTX_SIZE = 120;   // Größe der ProED_Context Struct

// --- Windows IOCTL Konstanten ---
const GENERIC_READ = 0x80000000;
const GENERIC_WRITE = 0x40000000;
const OPEN_EXISTING = 3;
const FILE_DEVICE_UNKNOWN = 0x00000022;
const METHOD_BUFFERED = 0;
const FILE_ANY_ACCESS = 0;

// IOCTL_PROKM_PROED = CTL_CODE(FILE_DEVICE_UNKNOWN, 0x801, METHOD_BUFFERED, FILE_ANY_ACCESS)
// Entspricht ((0x22 << 16) | (0 << 14) | (0x801 << 2) | 0) = 0x222004
const IOCTL_PROKM_PROED = 0x222004;

// --- LIBRARY DEFINITIONEN ---
const isWin = process.platform === 'win32';
const libName = isWin ? 'proed.dll' : 'libproed.so';

const proedLib = ffi.Library(libName, {
    'proed_get_version_info': ['string', ['pointer', 'pointer', 'pointer']],
    'proed_encrypt_envelope': ['int', ['pointer', 'pointer', size_t, 'pointer', size_t]],
    'proed_decrypt_envelope': ['int', ['pointer', 'pointer', size_t, 'pointer', size_t]]
});

// Windows Kernel32 API (nur laden, wenn wir auf Windows sind)
let kernel32 = null;
if (isWin) {
    kernel32 = ffi.Library('kernel32', {
        'CreateFileA': [pointer, ['string', uint32, uint32, pointer, uint32, uint32, pointer]],
        'DeviceIoControl': ['int', [pointer, uint32, pointer, uint32, pointer, uint32, pointer, pointer]],
        'CloseHandle': ['int', [pointer]]
    });
}

class ProED {
    /**
     * @param {Buffer} key 32-Byte Session Key
     */
    constructor(key) {
        if (!Buffer.isBuffer(key) || key.length !== KEY_SIZE) {
            throw new Error(`Key muss ein Buffer von exakt ${KEY_SIZE} Bytes sein.`);
        }
        this.key = key;
        this.hDevice = null;
        this.isKernelModeActive = false;

        // Prüfen, ob der Windows ProKM Kernel-Treiber erreichbar ist
        if (isWin && kernel32) {
            // GENERIC_READ | GENERIC_WRITE in JS Bitwise sicher machen (da JS Bitwise in 32-Bit signed ist)
            const access = (GENERIC_READ | GENERIC_WRITE) >>> 0;

            this.hDevice = kernel32.CreateFileA(
                "\\\\.\\ProKM", access, 0, null, OPEN_EXISTING, 0, null
            );

            // Pointer.address() gibt die Speicheradresse. -1 (0xFFFFFFFF) ist INVALID_HANDLE_VALUE
            if (!this.hDevice.isNull() && this.hDevice.address() !== -1) {
                this.isKernelModeActive = true;
            }
        }
    }

    static getVersion() {
        const major = ref.alloc(uint8);
        const minor = ref.alloc(uint8);
        const patch = ref.alloc(uint8);
        const info = proedLib.proed_get_version_info(major, minor, patch);
        return {
            info,
            version: `${major.deref()}.${minor.deref()}.${patch.deref()}`
        };
    }

    // =========================================================================
    // METHODE 1: KERNEL-MODE (Ring-0) - Raw High-Speed Verschlüsselung
    // =========================================================================
    processKernel(payload, iv, isDecrypt) {
        if (!this.isKernelModeActive) {
            throw new Error("ProKM Kernel-Driver ist nicht geladen. Fallback auf Software-Envelope erforderlich.");
        }
        if (!Buffer.isBuffer(iv) || iv.length !== 16) {
            throw new Error("IV muss exakt 16 Bytes lang sein.");
        }

        const payloadBuf = Buffer.isBuffer(payload) ? payload : Buffer.from(payload);
        const mode = isDecrypt ? 1 : 0;
        const dataLen = payloadBuf.length;

        // Struct Layout: Key(32) + IV(16) + Mode(4) + DataLen(4) + Payload(N)
        const reqSize = 32 + 16 + 4 + 4 + dataLen;
        const requestBuffer = Buffer.alloc(reqSize);

        this.key.copy(requestBuffer, 0);                 // [0-31] Key
        iv.copy(requestBuffer, 32);                      // [32-47] IV
        requestBuffer.writeUInt32LE(mode, 48);           // [48-51] Mode
        requestBuffer.writeUInt32LE(dataLen, 52);        // [52-55] DataLen
        payloadBuf.copy(requestBuffer, 56);              // [56-..] Payload

        const outBuffer = Buffer.alloc(reqSize);
        const bytesReturned = ref.alloc(uint32);

        // IOCTL Call an den Windows Kernel
        const success = kernel32.DeviceIoControl(
            this.hDevice, IOCTL_PROKM_PROED,
            requestBuffer, reqSize,
            outBuffer, reqSize,
            bytesReturned, null
        );

        if (success === 0) {
            throw new Error("Kernel DeviceIoControl fehlgeschlagen.");
        }

        // Die verarbeitete Payload beginnt im outBuffer ab Byte 56
        return outBuffer.slice(56, 56 + dataLen);
    }

    // =========================================================================
    // METHODE 2: SOFTWARE-MODE (Ring-3) - Armored Envelope (inkl. MAC)
    // =========================================================================
    encryptEnvelopeSoftware(payload) {
        const payloadBuf = Buffer.isBuffer(payload) ? payload : Buffer.from(payload);
        const totalLen = HEADER_SIZE + payloadBuf.length;

        const buffer = Buffer.alloc(totalLen);
        payloadBuf.copy(buffer, HEADER_SIZE); // Payload hinter den Header kopieren

        const ctx = Buffer.alloc(CTX_SIZE);
        const res = proedLib.proed_encrypt_envelope(ctx, this.key, this.key.length, buffer, totalLen);

        return res === 1 ? buffer : null;
    }

    decryptEnvelopeSoftware(packet) {
        if (!Buffer.isBuffer(packet) || packet.length < HEADER_SIZE) return null;

        const buffer = Buffer.from(packet);
        const ctx = Buffer.alloc(CTX_SIZE);

        const res = proedLib.proed_decrypt_envelope(ctx, this.key, this.key.length, buffer, buffer.length);

        if (res === 0) return null; // HMAC Check fehlgeschlagen!

        // Gibt nur die Payload (ohne Header) zurück
        return buffer.slice(HEADER_SIZE);
    }

    close() {
        if (this.isKernelModeActive && this.hDevice && kernel32) {
            kernel32.CloseHandle(this.hDevice);
            this.hDevice = null;
            this.isKernelModeActive = false;
        }
    }
}

module.exports = ProED;

File: ProED/Java/ProED.java
package com.brainai.proed;

import com.sun.jna.*;
import com.sun.jna.ptr.IntByReference;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Arrays;

public class ProED implements AutoCloseable {

    // --- KONSTANTEN ---
    public interface Config {
        int HEADER_SIZE = 48; // 16 Byte IV + 32 Byte HMAC
        int KEY_SIZE = 32;    // 256-Bit Key
        int CTX_SIZE = 120;   // Größe der ProED_Context Struct
        
        // --- Windows IOCTL Konstanten ---
        int GENERIC_READ = 0x80000000;
        int GENERIC_WRITE = 0x40000000;
        int OPEN_EXISTING = 3;
        int FILE_DEVICE_UNKNOWN = 0x00000022;
        int METHOD_BUFFERED = 0;
        int FILE_ANY_ACCESS = 0;
        
        // IOCTL_PROKM_PROED = CTL_CODE(FILE_DEVICE_UNKNOWN, 0x801, METHOD_BUFFERED, FILE_ANY_ACCESS)
        int IOCTL_PROKM_PROED = (FILE_DEVICE_UNKNOWN << 16) | (FILE_ANY_ACCESS << 14) | (0x801 << 2) | METHOD_BUFFERED;
    }

    // --- NATIVE INTERFACE: ProED DLL/SO (Software Fallback) ---
    public interface IProED extends Library {
        IProED INSTANCE = Native.load("proed", IProED.class);

        String proed_get_version_info(Pointer outMajor, Pointer outMinor, Pointer outPatch);
        int proed_encrypt_envelope(byte[] ctx, byte[] key, NativeSize keyLen, byte[] buffer, NativeSize totalLen);
        int proed_decrypt_envelope(byte[] ctx, byte[] key, NativeSize keyLen, byte[] buffer, NativeSize totalLen);
    }

    // --- NATIVE INTERFACE: Windows Kernel32 (Für Ring-0 IOCTL) ---
    public interface Kernel32 extends Library {
        Kernel32 INSTANCE = Platform.isWindows() ? Native.load("kernel32", Kernel32.class) : null;

        Pointer CreateFileA(String lpFileName, int dwDesiredAccess, int dwShareMode, Pointer lpSecurityAttributes, int dwCreationDisposition, int dwFlagsAndAttributes, Pointer hTemplateFile);
        boolean DeviceIoControl(Pointer hDevice, int dwIoControlCode, byte[] lpInBuffer, int nInBufferSize, byte[] lpOutBuffer, int nOutBufferSize, IntByReference lpBytesReturned, Pointer lpOverlapped);
        boolean CloseHandle(Pointer hObject);
    }

    // --- KLASSEN-LOGIK ---
    private final byte[] key;
    private Pointer hDevice = Pointer.NULL;
    private boolean isKernelModeActive = false;

    /**
     * Initialisiert den ProED Client und prüft, ob der Ring-0 Kernel-Treiber verfügbar ist.
     */
    public ProED(byte[] key) {
        if (key == null || key.length != Config.KEY_SIZE) {
            throw new IllegalArgumentException("Key muss exakt 32 Bytes lang sein.");
        }
        this.key = key;

        // Versuche, den Windows-Kernel-Treiber zu mounten (nur auf Windows)
        if (Platform.isWindows() && Kernel32.INSTANCE != null) {
            hDevice = Kernel32.INSTANCE.CreateFileA("\\\\.\\ProKM", 
                    Config.GENERIC_READ | Config.GENERIC_WRITE, 
                    0, Pointer.NULL, Config.OPEN_EXISTING, 0, Pointer.NULL);
                    
            // Pointer.createConstant(-1) entspricht INVALID_HANDLE_VALUE
            if (hDevice != null && Pointer.nativeValue(hDevice) != -1 && Pointer.nativeValue(hDevice) != 0) {
                isKernelModeActive = true;
            }
        }
    }

    public boolean isKernelModeActive() {
        return isKernelModeActive;
    }

    public String getVersionInfo() {
        Memory major = new Memory(1);
        Memory minor = new Memory(1);
        Memory patch = new Memory(1);
        return IProED.INSTANCE.proed_get_version_info(major, minor, patch);
    }

    // =========================================================================
    // METHODE 1: KERNEL-MODE (Ring-0) - Raw High-Speed Verschlüsselung
    // =========================================================================
    public byte[] processKernel(byte[] payload, byte[] iv, boolean isDecrypt) {
        if (!isKernelModeActive) {
            throw new IllegalStateException("ProKM Kernel-Driver ist nicht geladen. Fallback auf Software-Envelope erforderlich.");
        }
        if (iv == null || iv.length != 16) {
            throw new IllegalArgumentException("IV muss exakt 16 Bytes lang sein.");
        }

        int mode = isDecrypt ? 1 : 0;
        int dataLen = payload.length;

        // Struct Layout: Key(32) + IV(16) + Mode(4) + DataLen(4) + Payload(N)
        int reqSize = 32 + 16 + 4 + 4 + payload.length;
        byte[] requestBuffer = new byte[reqSize];

        // Speicher zusammenbauen (Little Endian für die Integer-Werte)
        ByteBuffer bb = ByteBuffer.wrap(requestBuffer).order(ByteOrder.LITTLE_ENDIAN);
        bb.put(key);
        bb.put(iv);
        bb.putInt(mode);
        bb.putInt(dataLen);
        bb.put(payload);

        byte[] outBuffer = new byte[reqSize];
        IntByReference bytesReturned = new IntByReference();

        // IOCTL Call an den Windows Kernel
        boolean success = Kernel32.INSTANCE.DeviceIoControl(
                hDevice, Config.IOCTL_PROKM_PROED,
                requestBuffer, reqSize,
                outBuffer, reqSize,
                bytesReturned, Pointer.NULL);

        if (!success) {
            int err = Native.getLastError();
            throw new RuntimeException("Kernel DeviceIoControl fehlgeschlagen. Windows Error Code: " + err);
        }

        // Die verarbeitete Payload beginnt im outBuffer ab Byte 56
        byte[] result = new byte[payload.length];
        System.arraycopy(outBuffer, 56, result, 0, payload.length);

        return result;
    }

    // =========================================================================
    // METHODE 2: SOFTWARE-MODE (Ring-3) - Armored Envelope (inkl. MAC)
    // =========================================================================
    public byte[] encryptEnvelopeSoftware(byte[] payload) {
        if (payload == null) throw new IllegalArgumentException("Payload darf nicht null sein.");
        
        byte[] buffer = new byte[Config.HEADER_SIZE + payload.length];
        System.arraycopy(payload, 0, buffer, Config.HEADER_SIZE, payload.length);
        
        byte[] ctx = new byte[Config.CTX_SIZE];
        int res = IProED.INSTANCE.proed_encrypt_envelope(
                ctx, key, new NativeSize(key.length), buffer, new NativeSize(buffer.length));
        
        return (res == 1) ? buffer : null;
    }

    public byte[] decryptEnvelopeSoftware(byte[] packet) {
        if (packet == null || packet.length < Config.HEADER_SIZE) return null;
        
        byte[] buffer = Arrays.copyOf(packet, packet.length);
        byte[] ctx = new byte[Config.CTX_SIZE];
        
        int res = IProED.INSTANCE.proed_decrypt_envelope(
                ctx, key, new NativeSize(key.length), buffer, new NativeSize(buffer.length));
        
        if (res == 0) return null; // MAC Error (Manipulation erkannt!)
        
        byte[] payload = new byte[buffer.length - Config.HEADER_SIZE];
        System.arraycopy(buffer, Config.HEADER_SIZE, payload, 0, payload.length);
        
        return payload;
    }

    @Override
    public void close() {
        if (isKernelModeActive && hDevice != null) {
            Kernel32.INSTANCE.CloseHandle(hDevice);
            hDevice = Pointer.NULL;
            isKernelModeActive = false;
        }
    }
}